src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -539,7 +546,101 @@ void QgsSymbolLegendNode::updateLabel()
   emit dataChanged();
 }
 
+QString QgsSymbolLegendNode::evaluateLabel( QgsExpressionContext context, QString label )
+{
+  if ( !mLayerNode )
+    return QString();
+
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  if ( label.isEmpty() )
+  {
+    if ( mEmbeddedInParent )
+    {
+      QString layerName = mLayerNode->name();
+      if ( !mLayerNode->customProperty( QStringLiteral( "legend/title-label" ) ).isNull() )
+        layerName = mLayerNode->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+
+      mLabel = mUserLabel.isEmpty() ? layerName : mUserLabel;
+      if ( vl && ( !( mLayerNode->expression().isEmpty() ) || mLabel.contains( "[%" ) ) )
+      {
+        mLabel = evaluateLabelExpression( mLabel, vl, context );
+      }
+    }
+    else
+    {
+      mLabel = mUserLabel.isEmpty() ? mItem.label() : mUserLabel;
+      if ( vl && ( !( mLayerNode->expression().isEmpty() ) || mLabel.contains( "[%" ) ) )
+      {
+        mLabel = evaluateLabelExpression( mLabel, vl, context );
+      }
+    }
+    emit dataChanged();
+    return mLabel;
+  }
+  else
+  {
+    if ( vl && ( !( mLayerNode->expression().isEmpty() ) || label.contains( "[%" ) ) )
+    {
+      label = evaluateLabelExpression( label, vl, context );
+    }
+    return label;
+  }
+}
+
+QgsExpressionContext QgsSymbolLegendNode::createExpressionContext( QgsExpressionContext context ) const
+{
+  //= QgsLayoutItem::createExpressionContext();
+
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  // TODO: ADD NECESSARY SCOPES ??
+  context.appendScope( vl->createExpressionContextScope() );
+
+
+  QgsExpressionContextScope *scope = new QgsExpressionContextScope( tr( "Symbol scope" ) );
+
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_label" ), textOnSymbolLabel(), true ) );
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "symbol_id" ), mItem.ruleKey(), true ) );
+  QgsVectorLayerFeatureCounter *counter = vl->countSymbolFeatures( true );
```

Comment:
Isn't this reducing performance a lot by blocking the UI (even for every single symbol)?
Would it be possible to just use the current counter results (and add a substitution while not available) and update the label when the counter job has finished in the background?

Reply:
Yes, but I am not sure how to implement such a thing to keep the counting going on and postpone the evaluation.

If you have a suggestion please let me know.

Reply:
What I can think about is pushing the whole function in QgsApplication::taskManager()->addTask( mFeatureCounter ); and still using the wait,

Or finding another way to delay the evaluation and resume the code execution only when[ the signal from the counter ](https://github.com/roya0045/QGIS/blob/595c9392536d6a17e316280d29f6542b7e4508f0/src/core/qgsvectorlayer.cpp#L725) is emitted. And to maintaining all this is the background to not stall QGIS.

The first option should be somewhat straighforward (I doubt I need to add signal but please correct me) the second one I'm not sure if it's doable. @nyalldawson  any opinion on this?

Reply:
Signals were used, I'll mark this as resolved